File: matiasmasso/Mat.NET/DT2/DTOSpv.vb
Public Class DTOSpv
    Inherits DTOBaseGuid

    Public Shadows Property Guid As Guid
        'obligatori per serializer
        Get
            Return MyBase.Guid
        End Get
        Set(value As Guid)
            MyBase.Guid = value
        End Set
    End Property

    Property Emp As DTOEmp
    Property Id As Integer
    Property FchAvis As Date
    Property FchRead As Date
    Property Customer As DTOCustomer
    Property SolicitaGarantia As Boolean
    Property Garantia As Boolean
    Property Contacto As String
    Property sRef As String
    Property SpvIn As DTOSpvIn
    Property Product As Object
    Property SerialNumber As String
    Property ObsClient As String
    Property ObsTecnic As String
    Property LabelEmailedTo As String
    Property Nom As String
    Property Address As DTOAddress
    Property Tel As String

    Property ValJob As DTOAmt
    Property ValMaterial As DTOAmt
    Property ValEmbalatje As DTOAmt
    Property ValPorts As DTOAmt
    Property Delivery As DTODelivery
    Property Incidencia As DTOIncidencia

    Property UsrRegister As DTOUser
    Property UsrTecnic As DTOUser

    Property UsrOutOfSpvIn As DTOUser
    Property FchOutOfSpvIn As Date
    Property ObsOutOfSpvIn As String
    Property UsrOutOfSpvOut As DTOUser
    Property FchOutOfSpvOut As Date
    Property ObsOutOfSpvOut As String

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function Factory(oUsrRegister As DTOUser) As DTOSpv
        Dim retval As New DTOSpv
        With retval
            .Emp = oUsrRegister.Emp
            .UsrRegister = oUsrRegister
            .ValJob = DTOAmt.Empty()
            .FchAvis = Today
        End With
        Return retval
    End Function

    Public Function GetProduct() As DTOProduct
        Dim retval As DTOProduct = Nothing
        If _Product IsNot Nothing Then
            If _Product.GetType().IsSubclassOf(GetType(DTOProduct)) Then
                retval = _Product
            Else
                Dim oProduct As DTOProduct = _Product.toobject(Of DTOProduct)
                Select Case oProduct.SourceCod
                    Case DTOProduct.SourceCods.SKU
                        retval = _Product.toobject(Of DTOProductSku)
                    Case DTOProduct.SourceCods.Category
                        retval = _Product.toobject(Of DTOProductCategory)
                    Case DTOProduct.SourceCods.Brand
                        retval = _Product.toobject(Of DTOProductBrand)
                End Select
            End If
        End If
        Return retval
    End Function

    Public Function Lines(oLang As DTOLang) As List(Of String)
        Dim retval As New List(Of String)

        Dim Str As String = oLang.tradueix("Reparación num.", "Reparació num.", "Service Job num.")
        Str = Str & _Id & " "
        If _Garantia Then
            Str = Str & oLang.tradueix("en garantía", "en garantía", "under warranty")
        Else
            Str = Str & oLang.tradueix("con cargo", "amb carrec", "")
        End If
        retval.Add(Str)

        Str = oLang.tradueix("Solicitada", "Sol·licitada", "Requested")
        If _Contacto > "" Then
            Str = Str & oLang.tradueix(" por ", " per ", " by ") & _Contacto
        End If
        Str = Str & oLang.tradueix(" en fecha ", " en data ", " on ") & _FchAvis.ToShortDateString
        retval.Add(Str)

        Str = oLang.tradueix("Recibida el ", "Rebuda el ", "Received on ") & _SpvIn.Fch.ToShortDateString
        If _sRef > "" Then
            Str = Str & " (s/ref.: " & _sRef & ")"
        End If
        retval.Add(Str)

        Str = oLang.tradueix("Producto: ", "Producte: ", "Product: ") & DTOProduct.GetNom( _Product)
        retval.Add(Str)

        If _SerialNumber > "" Then
            Str = oLang.tradueix("Número de serie: ", "Número de serie: ", "Serial number: ") & _SerialNumber
            retval.Add(Str)
        End If

        retval.Add("")

        Dim sObsTecnic As String() = _ObsTecnic.Split(vbLf)
        For Each Str In sObsTecnic
            retval.Add(Str.Trim)
        Next

        retval.Add("")

        Return retval
    End Function

    Shared Function TextRegistre(oSpv As DTOSpv, ByVal oLang As DTOLang) As String
        Dim sUsrRegisterText As String = ""
        Dim ESP As String = "registrado"
        Dim CAT As String = "registrat"
        Dim ENG As String = "recorder"

        If oSpv.UsrRegister IsNot Nothing Then
            Dim sLogin As String = DTOUser.NicknameOrElse(oSpv.UsrRegister)
            ESP = ESP & " por " & sLogin
            CAT = CAT & " per " & sLogin
            ENG = ENG & " by " & sLogin
        End If

        Dim sFch As String = oSpv.FchAvis.ToShortDateString
        ESP = ESP & " el " & sFch
        CAT = CAT & " el " & sFch
        ENG = ENG & " on " & sFch

        Dim retVal As String = oLang.tradueix(ESP, CAT, ENG)
        Return retVal
    End Function

    Shared Function textOutSpvIn(oSpv As DTOSpv) As String
        Dim s As String = "retirat de pendents de entrar per " & DTOUser.NicknameOrElse(oSpv.UsrOutOfSpvIn)
        s = s & " el " & oSpv.FchOutOfSpvIn.ToShortDateString
        If oSpv.ObsOutOfSpvIn > "" Then
            s = s & ": " & oSpv.ObsOutOfSpvIn
        End If
        Return s
    End Function

    Shared Function textOutSpvOut(oSpv As DTOSpv) As String
        Dim s As String = "retirat de pendents de sortir per " & DTOUser.NicknameOrElse(oSpv.UsrOutOfSpvOut)
        s = s & " el " & oSpv.FchOutOfSpvOut.ToShortDateString
        If oSpv.ObsOutOfSpvOut > "" Then
            s = s & ": " & oSpv.ObsOutOfSpvOut
        End If
        Return s
    End Function

End Class

File: matiasmasso/Mat.NET/DTO/DTOSpv.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOSpv : DTOBaseGuid
    {
        public new Guid Guid
        {
            // obligatori per serializer
            get
            {
                return base.Guid;
            }
            set
            {
                base.Guid = value;
            }
        }

        public DTOEmp emp { get; set; }
        public int id { get; set; }
        public DateTime fchAvis { get; set; }
        public DateTime fchRead { get; set; }
        public DTOCustomer customer { get; set; }
        public bool solicitaGarantia { get; set; }
        public bool garantia { get; set; }
        public string contacto { get; set; }
        public string sRef { get; set; }
        public DTOSpvIn spvIn { get; set; }
        public object product { get; set; }
        public string serialNumber { get; set; }

        public string ManufactureDate { get; set; }
        public string obsClient { get; set; }
        public string obsTecnic { get; set; }
        public string labelEmailedTo { get; set; }
        public string nom { get; set; }
        public DTOAddress address { get; set; }
        public string tel { get; set; }

        public DTOAmt valJob { get; set; }
        public DTOAmt valMaterial { get; set; }
        public DTOAmt valEmbalatje { get; set; }
        public DTOAmt valPorts { get; set; }
        public DTODelivery delivery { get; set; }
        public DTOIncidencia incidencia { get; set; }

        public DTOUser usrRegister { get; set; }
        public DTOUser usrTecnic { get; set; }

        public DTOUser usrOutOfSpvIn { get; set; }
        public DateTime fchOutOfSpvIn { get; set; }
        public string obsOutOfSpvIn { get; set; }
        public DTOUser usrOutOfSpvOut { get; set; }
        public DateTime fchOutOfSpvOut { get; set; }
        public string obsOutOfSpvOut { get; set; }

        public DTOSpv() : base()
        {
        }

        public DTOSpv(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOSpv Factory(DTOUser oUsrRegister)
        {
            DTOSpv retval = new DTOSpv();
            {
                var withBlock = retval;
                withBlock.emp = oUsrRegister.Emp;
                withBlock.usrRegister = oUsrRegister;
                withBlock.valJob = DTOAmt.Empty();
                withBlock.fchAvis = DTO.GlobalVariables.Today();
            }
            return retval;
        }


        public void restoreObjects()
        {
            this.product = DTOProduct.fromJObject(this.product as JObject);

            if (this.incidencia != null)
                this.incidencia.restoreObjects();
        }


        public List<string> lines(DTOLang oLang)
        {
            List<string> retval = new List<string>();

            string Str = oLang.Tradueix("Reparación num.", "Reparació num.", "Service Job num.");
            Str = Str + id + " ";
            if (garantia)
                Str = Str + oLang.Tradueix("en garantía", "en garantía", "under warranty");
            else
                Str = Str + oLang.Tradueix("con cargo", "amb carrec", "");
            retval.Add(Str);

            Str = oLang.Tradueix("Solicitada", "Sol·licitada", "Requested");
            if (contacto.isNotEmpty())
                Str = Str + oLang.Tradueix(" por ", " per ", " by ") + contacto;
            Str = Str + oLang.Tradueix(" en fecha ", " en data ", " on ") + fchAvis.ToShortDateString();
            retval.Add(Str);

            Str = oLang.Tradueix("Recibida el ", "Rebuda el ", "Received on ") + spvIn.fch.ToShortDateString();
            if (sRef.isNotEmpty())
                Str = Str + " (s/ref.: " + sRef + ")";
            retval.Add(Str);

            if (incidencia != null)
            {
                Str = oLang.Tradueix("Incidencia ", "Incidència ", "Incidence ") + incidencia.AsinOrNum() + " del " + incidencia.Fch.ToShortDateString();
                retval.Add(Str);
            }

            DTOProduct oProduct = null;
            if (product is JObject)
                oProduct = DTOProduct.fromJObject((JObject)product);
            else
                oProduct = (DTOProduct)product;

            Str = oLang.Tradueix("Producto: ", "Producte: ", "Product: ") + oProduct.FullNom(oLang);
            retval.Add(Str);

            if (serialNumber.isNotEmpty())
            {
                Str = oLang.Tradueix("Número de serie: ", "Número de serie: ", "Serial number: ", "Número de serie: ") + serialNumber;
                retval.Add(Str);
            }

            if (ManufactureDate.isNotEmpty())
            {
                Str = oLang.Tradueix("Fecha de fabricación: ", "Data de fabricació: ", "Manufacturing date: ") + ManufactureDate;
                retval.Add(Str);
            }

            retval.Add("");

            List<string> sObsTecnic = obsTecnic.toLinesList();
            foreach (string str in sObsTecnic)
                retval.Add(str.Trim());

            retval.Add("");

            return retval;
        }

        public static string TextRegistre(DTOSpv oSpv, DTOLang oLang)
        {
            string ESP = "registrado";
            string CAT = "registrat";
            string ENG = "recorder";

            if (oSpv.usrRegister != null)
            {
                string sLogin = DTOUser.NicknameOrElse(oSpv.usrRegister);
                ESP = ESP + " por " + sLogin;
                CAT = CAT + " per " + sLogin;
                ENG = ENG + " by " + sLogin;
            }

            string sFch = oSpv.fchAvis.ToShortDateString();
            ESP = ESP + " el " + sFch;
            CAT = CAT + " el " + sFch;
            ENG = ENG + " on " + sFch;

            string retVal = oLang.Tradueix(ESP, CAT, ENG);
            return retVal;
        }

        public static string textOutSpvIn(DTOSpv oSpv)
        {
            string s = "retirat de pendents de entrar per " + DTOUser.NicknameOrElse(oSpv.usrOutOfSpvIn);
            s = s + " el " + oSpv.fchOutOfSpvIn.ToShortDateString();
            if (oSpv.obsOutOfSpvIn.isNotEmpty())
                s = s + ": " + oSpv.obsOutOfSpvIn;
            return s;
        }

        public static string textOutSpvOut(DTOSpv oSpv)
        {
            string s = "retirat de pendents de sortir per " + DTOUser.NicknameOrElse(oSpv.usrOutOfSpvOut);
            s = s + " el " + oSpv.fchOutOfSpvOut.ToShortDateString();
            if (oSpv.obsOutOfSpvOut.isNotEmpty())
                s = s + ": " + oSpv.obsOutOfSpvOut;
            return s;
        }
    }
}
